### DISCUSSION TITLE
Options for Document.transform to prevent empty node removal

### DISCUSSION THREAD

**Discussion by ReedCopsey** [donmccurdy/glTF-Transform#1050] [General]
Hello,

I am currently using Document.transform() to add draco compression to an existing GLB.

However, in the process, this strips out scene nodes from the GLB file which are empty (other than metadata). For our custom application, we rely on the existence of these for various features, as the metadata can provide instructions for things that happen post-load of the GLB.

Is there a way to run this through Document.transform to draco compress without stripping out empty nodes? Currently, we only call this with draco(), ie:

```
        await glbDocument.transform(
            draco()
        );
```

I've also tried explicitly adding prune() with options for keeping leaves or attributes, as they seemed potentially related, but that has no impact - the empty scene nodes are still removed.

Thank you.

**Comment by donmccurdy** [comment]
Hi @ReedCopsey —

What's happening here is that Draco depends on having indexed geometry. The `draco()` transform calls `weld({tolerance=0})` to losslessly weld, and ensure there's an index on every geometry. In some cases (where tolerance>0) welding could remove meshes, and so that transform also looks for empty nodes that might have been left behind, and removes them. Needlessly, in this case. 

I think we can fix all that to not cause the issue, yes. In the meantime, if you know your geometry is already indexed, you can run a simplified version of the Draco transform:

```javascript
import { KHRDracoMeshCompression } from '@gltf-transform/extensions';

const dracoBasic = () => {
  return (document) => {
    document
      .createExtension(KHRDracoMeshCompression)
      .setRequired(true)
  };
};

await document.transform(dracoBasic());
```

**Comment by ReedCopsey** [comment]
Thanks @donmccurdy - That works for me, at least for now!

Would be great to have an option to avoid the prune call in the normal draco extensions, so welding could still occur. I'm not sure how weld with a tolerance of 0 could ever remove a mesh, so in the case of draco(), it seems like this shouldn't prune at all. Am I missing something?

**Reply by donmccurdy** [reply]
>  I'm not sure how weld with a tolerance of 0 could ever remove a mesh...

You're correct, it can't, but pruning isn't (yet) checking the tolerance level. That can be fixed, when I have the time, or feel free to make a PR here:

https://github.com/donmccurdy/glTF-Transform/blob/8fcc53cfd29649aa3174997fe17138b47897db4a/packages/functions/src/weld.ts#L114-L136

**Comment by ReedCopsey** [comment]
Thanks - Submitted a PR here - https://github.com/donmccurdy/glTF-Transform/pull/1059

I think this is all that's required for this? (It's just a couple of lines of code wrapping that prune call). Let me know if you want any other tweaks to it, though. This would be better than the simplified draco option (which I'm using now), as that doesn't support all of the options the normal draco function supports.
